open-metadata-implementation/access-services/cognos/cognos-server/src/main/java/org/odpi/openmetadata/accessservices/cognos/assets/DatabaseContextHandler.java

```diff
@@ -0,0 +1,500 @@
+	 */
+	void setContext() {
+		omEntityDao.setContext(new Throwable().getStackTrace()[1].getMethodName());
+	}
+
+
+	/**
+	 * Get list of databases on the server.
+	 * @return list of database descriptors.
+	 */
+	public List<ResponseContainerDatabase> getDatabases() {
+		
+		setContext();
+
+		InstanceProperties instanceProperties = omEntityDao.buildMatchingInstanceProperties(Collections.emptyMap(), true);
+		List<EntityDetail> entities = omEntityDao.findEntities(instanceProperties, Constants.DATABASE, 0, 0);
+		return Optional.ofNullable(entities).map(Collection::stream).orElseGet(Stream::empty)
+				.parallel()
+				.map(this::buildDatabase)
+				.filter(Objects::nonNull).collect(Collectors.toList());
+	}
+	
+	private ResponseContainerDatabase buildDatabase(EntityDetail e) {
+		ResponseContainerDatabase ret = new ResponseContainerDatabase();
+		ret.setDbName(this.getEntityStringProperty(e, Constants.NAME));
+		ret.setDbType(this.getEntityStringProperty(e, Constants.TYPE));
+		ret.setDbVersion(this.getEntityStringProperty(e, Constants.VERSION));
+		ret.setGUID(e.getGUID());
+		return ret;
+	}
+
+
+
+	/**
+	 * Retrieve schemas for given database from repository.
+	 * 
+	 * @param guidDataSource defines database
+	 * @return list of schemas attributes.
+	 */
+	public List<ResponseContainerDatabaseSchema> getDatabaseSchemas(String guidDataSource) {
+
+		setContext();
+
+		EntityDetail db = omEntityDao.getEntityByGuid(guidDataSource);
+		String catalogName = getEntityStringProperty(db, Constants.ATTRIBUTE_NAME);
+
+		List<Relationship> db2SchemaRelationships = omEntityDao.getRelationshipsForEntity(db, Constants.DATA_CONTENT_FOR_DATASET);
+
+		if (db2SchemaRelationships == null) {
+			return Collections.emptyList();	// no schemas
+		}
+		
+		List<ResponseContainerDatabaseSchema> ret = db2SchemaRelationships.parallelStream()
+				.map(this::getSchemaEntityFromRelationship)
+				.filter(Objects::nonNull)
+				.map(e->buildSchemaForRelationship(catalogName, e))
+				.filter(Objects::nonNull).collect(Collectors.toList());
+		
+		ret.sort(Comparator.comparing(e->e.getSchema()));
+
+		return ret;
+	}
+
+	private ResponseContainerDatabaseSchema buildSchemaForRelationship(String catalogName, EntityDetail dbSchemaEntity) {
+			String schemaName = getEntityStringProperty(dbSchemaEntity, Constants.ATTRIBUTE_NAME);
+			ResponseContainerDatabaseSchema schema = new ResponseContainerDatabaseSchema();
+			schema.setCatalog(catalogName);
+			schema.setSchema(schemaName);
+			schema.setId(schema.buildId());
+		return schema;
+	}
+
+	/**
+	 * Get schema entity from relationship catalog->schema
+	 * @param r relationship
+	 * @return schema entity.
+	 */
+	private EntityDetail getSchemaEntityFromRelationship(Relationship r) {
+		try {
+			return omEntityDao.getEntityByGuid(r.getEntityTwoProxy().getGUID());
+		} catch (CognosRuntimeException e) {
+			// don't throw 
+			log.warn(e.getMessage(), e);
```

Comment:
Debug logging should not be used for error handling.  This logging will be turned off in a production environment.

We also avoid using variable names of a single letter because it makes it harder for someone else to read.